src/construction/construction.py

```diff
@@ -1,16 +1,243 @@
+    amendments_exist = check_file_exists(amendments_filepath)
+    additions_exist = check_file_exists(additions_filepath)
+
+    # If each file exists, read it and call the function to apply them
+    if (not amendments_exist) and (not additions_exist):
+        construction_logger.info("No amendments or additions to apply, skipping...")
+        return main_df
+
+    if amendments_exist:
+        try:
+            amendments_df = read_csv(amendments_filepath)
+            constructed_df = apply_amendments(main_df, amendments_df)
+        except pd.errors.EmptyDataError:
+            construction_logger.warning(f"Amendments file {amendments_filepath} is empty, skipping...")
+
+    if additions_exist:
+        try:
+            additions_df = read_csv(additions_filepath)
+            constructed_df = apply_additions(main_df, additions_df)
+        except pd.errors.EmptyDataError:
+            construction_logger.warning(f"Additions file {additions_filepath} is empty, skipping...")
+
+    # Save the constructed dataframe as a CSV
+    tdate = datetime.now().strftime("%Y-%m-%d")
+    # ? Should this go in a subfolder of the outputs folder?
+    construction_output_filepath = os.path.join(paths["output_path"], f"construction_amendments_{tdate}_v{run_id}.csv")
+    write_csv(construction_output_filepath, constructed_df)
+    return constructed_df
+
+
+def apply_amendments(main_df, amendments_df):
+    """Apply amendments to the main snapshot."""
+    key_cols = ["reference", "year", "instance"]
+    numeric_cols = ["219", "220", "242", "243", "244", "245",
+                    "246", "247", "248", "249", "250"]
+    numeric_cols_new = [f"{i}_updated" for i in numeric_cols]
+
+    accepted_amendments_df = amendments_df.drop(amendments_df[~amendments_df.accept_changes])
+
+    if accepted_amendments_df.shape[0] == 0:
+        construction_logger.info("No amended records found during construction")
+        return main_df
+
+    # Drop the diff columns
+    accepted_amendments_df = accepted_amendments_df.drop([col for col in accepted_amendments_df.columns if col.endswith("_diff")])
+
+    # Join the amendments onto the main snapshot
+    amended_df = pd.merge(main_df, accepted_amendments_df,
+                          how="left",
+                          on=key_cols)
+
+    # Drop the old numeric cols and rename the amended cols
+    amended_df = amended_df.drop(numeric_cols)
+    cols_to_rename = dict(zip(numeric_cols_new, numeric_cols))
+    amended_df = amended_df.rename(columns=cols_to_rename, errors="raise")
+
+    construction_logger.info(f"{accepted_amendments_df.shape[0]} records amended during construction")
+
+    return amended_df
+
+
+def apply_additions(main_df, additions_df):
+    """Apply additions to the main snapshot."""
+    # Drop records where accept_changes is False and if any remain, add them to main df
+    accepted_additions_df = additions_df.drop(additions_df[~additions_df.accept_changes])
```

Comment:
Getting a KeyError here: think it needs accepted_additions_df = additions_df.drop(additions_df[~additions_df.accept_changes]**, axis=1**)

Reply:
Ignore the above!  Just relooked at this, and that wasn't the problem - it was dropping all the records, even if accept_changes was set to TRUE.  Tweaked it to: accepted_additions_df = additions_df.drop(additions_df[~additions_df["accept_changes"]].index) and it has run and the additional records are showing in the final constructed df

Reply:
Did you mark one of the first few rows with `accept_changes` true? The first few records don't have anything in the 202 column and it looks like the zero division is happening with an operation in 202 as well.

Reply:
Ah, yes - I just marked the first 13 as true

Reply:
I had everything pointed at R:\BERD Results System Development 2023\DAP_emulation\construction, so you can see my outputs (construction_final_2023-11-03_v74 is the constructed df at the end of construction with additional records - I used construction_additions_2023-11-02_v50  as my additions file, with top 13 records marked as true)

Reply:
Just ran it and it gives the same error even when they both have 202s. But the `calc_fte_column()` function is using a list of new and old columns for its loop - I know there's some differences between the columns in the 2022 data and the 2020 snapshot I'm using as the fake updated snapshot, so it might be that? It's hard to diagnose since I don't know what that function is for!

Reply:
After a bunch of testing I can't find any reason for this. The function divides by `tot_202_CD` but this column is nonzero when the error occurs. At the moment I'm fairly confident this is an issue with the fake updated snapshot I have to use outside of DAP.